0002-crypto-conditions/0002-crypto-conditions.md

```diff
@@ -114,312 +114,288 @@ This specification is a part of the [Interledger Protocol](https://interledger.o
+This is the maximum length of the essential fulfillment payload for a fulfillment that can fulfill this condition.
 
-Feature suites are chosen such that they represent reasonable clusters of functionality. For instance, it is reasonable to require that an implementation which supports SHA-256 in one context MUST support it in all contexts, since it already needed to implement the algorithm.
+For each crypto-condition type, a formula is provided for calculating the maximum fulfillment length to ensure that implementations produce consistent output.
 
-An implementation which supports a certain set of feature suites MUST accept all condition types which depend only on that set or any subset of feature suites.
+When a crypto-condition is submitted to an implementation, this implementation MUST verify that it will be able to process a fulfillment with a payload of size maxFulfillmentLength plus any additional encoding bytes.
 
-| Suite Name | Feature Bit | Feature Bit (BASE16) | Summary |
-|------------|-----|------|---------------------------------|
-| SHA-256    | 2^0 | 0x01 | The SHA-256 hashing algorithm. |
-| PREIMAGE   | 2^1 | 0x02 | The functionality of comparing a hash to a preimage. |
-| PREFIX     | 2^2 | 0x04 | The functionality of prefixing the fulfillment with a prefix before generating the condition. |
-| THRESHOLD  | 2^3 | 0x08 | The functionality of composing a condition out of several weighted subconditions. |
-| RSA-PSS    | 2^4 | 0x10 | The RSA-PSS signature algorithm. |
-| ED25519    | 2^5 | 0x20 | The ED25519 signature algorithm. |
+### Subtypes
 
-## SHA-256 {#sha-256-feature-suite}
+It must be possible to verify that all types used in a crypto-condition are supported (including the types and subtypes of any sub-crypto-conditions) even if the fulfillment is not available yet. Therefore, all compound conditions contain a bitmask to indicate the set of types and subtypes of all sub-crypto-conditions. The subtypes field is only present in the compound conditions (THRESHOLD-SHA-256 and PREFIX-SHA-256). 
 
-The SHA-256 feature suite provides the SHA-256 hash function. SHA-256 is a cryptographic hash function published by the US National Institute of Standards and Technology that produces 256 bits of output. This feature suite is assigned the feature bit 2^0 = 0x01.
+The field is encoded as a variable length octet string and specifies the set of types an implementation must support in order to be able to successfully validate the fulfillment of this condition. This is the set of types and subtypes of the top-level condition and all sub-crypto-conditions, recursively. 
 
-## PREIMAGE {#preimage-feature-suite}
-The PREIMAGE feature suite provides conditions that use a preimage as a one-time signature. This feature suite is assigned the feature bit 2^1 = 0x02.
+Each bit represents a type. The list of known types is the IANA-maintained [Crypto-Condition Type Registry](#crypto-conditions-type-registry) .
```

Comment:
So, a number of issues.

First of all, I don't see where we define the types for the existing 5 cryptoconditions. We can't use their types, since those are an `ENUM` and can't be used as a bitfield, at least as defined since some types have more than 1 bit set.

If we keep `type` as an `ENUM` and make the enum members have values usable as bitfields we're limited to 8 condition types (or 16, 24, 32, etc). Obviously not the best idea.

If we keep type as is, but want `subtypes` to be a bitfield we need to "duplicate" each type in the `ENUM` with a "subtype" whose value is a power of 2; that just seems error-prone and unnecessarily complex.

Also, if we are going to treat these values as a bitfield, then we should probably encode them as such: using an OER "bit string" without a size constraint. Not as `INTEGER`.

I think that the existing `type` + `feature` setup works well. It also allows us to easily "reuse" a type in a safe way. For example, want a `PREFIX-SHA-512`? Don't define a new type; just use the existing `PREFIX` type with `SHA-512` as the feature suite.